File: bin/validate_audio.php
<?php
require_once(__DIR__ . '/../include/constants.php');
require_once(__DIR__ . '/../include/dbconnect.php');

$workerId = 'worker_' . getmypid();

while (true) {
    $mysqli->autocommit(false);
    
    $stmt = $mysqli->prepare("SELECT a.id, s.textline, a.clip_path FROM audio_clip as a, statements as s WHERE status = ? AND model = 'cleaned' AND a.statement_id = s.id LIMIT 1");
    $stmt->bind_param('i', $STATUS_NEW);
    $stmt->execute();
    $result = $stmt->get_result();
    $textclip = $result->fetch_all(MYSQLI_ASSOC);

    if (empty($textclip)) {
        $mysqli->rollback();
        echo "No more pending clips. Worker $workerId is done.\n";
        break;
    }

    $currentClip = $textclip[0];
 
    $stmt = $mysqli->prepare('UPDATE audio_clip SET status = ? WHERE id = ?;');
    $stmt->bind_param("ii", 
        $STATUS_PENDING,
        $currentClip['id']
    );
    $stmt->execute();

    $mysqli->commit();

    $clipId = $currentClip['id'];
    echo "Worker $workerId processing clip #$clipId...\n";
    handleClipRecord($currentClip);
}

function cleanString($str) {
    $str = preg_replace("/[^\p{L}\s\d]+/u", "", $str);
    $str = strtolower(trim($str));
    $str = preg_replace("/\s+/u", " ", $str);
    return $str;
}

function handleClipRecord($currentClip) {
    global $mysqli, $REPO_DIR, $STATUS_MACHINE_APPROVED, $STATUS_DO_USER_REVIEW;
    $cmdSilence = sprintf(
        'whisper --language Swedish -f txt %s 2>&1',
        escapeshellarg($REPO_DIR . $currentClip['clip_path']),
    );
    $outputSoX = shell_exec($cmdSilence);

    $textFile = basename($currentClip['clip_path'], '.mp3') . '.txt'; 
    if (file_exists($textFile)) {
        $cleanedTextLine = cleanString($currentClip['textline']);
        $cleanedTextFile = cleanString(file_get_contents($textFile));

        if (strcmp($cleanedTextFile, $cleanedTextLine) == 0) {
            echo "Approved : " . $cleanedTextFile . "\n";
            $stmt = $mysqli->prepare('UPDATE audio_clip SET status = ? WHERE id = ?;');
            $stmt->bind_param("ii", 
                $STATUS_MACHINE_APPROVED,
                $currentClip['id']
            );
            $stmt->execute();
        } else {
            echo "User required : " . $cleanedTextFile . "\n";
            $stmt = $mysqli->prepare('UPDATE audio_clip SET status = ? WHERE id = ?;');
            $stmt->bind_param("ii", 
                $STATUS_DO_USER_REVIEW,
                $currentClip['id']
            );
            $stmt->execute();
        }
        unlink($textFile);
    }
}

File: bin/validate_audio.py
import whisper
import mysql.connector
import time
import json
import regex

def clean_string(s: str) -> str:
    s = regex.sub(r'[^\p{Letter}\s\d]+', '', s)
    s = s.strip().lower()
    s = regex.sub(r'\s+', ' ', s)
    return s

with open("../etc/mysql_config.json", "r") as f:
    config_data = json.load(f)

conn = mysql.connector.connect(
    host=config_data["mysql"]["host"],
    user=config_data["mysql"]["user"],
    password=config_data["mysql"]["password"],
    database=config_data["mysql"]["database"],
)
cursor = conn.cursor()

cursor.execute("SELECT a.id, s.textline, a.clip_path FROM audio_clip as a, statements as s WHERE status = 0 AND model = 'cleaned' AND a.statement_id = s.id")
rows = cursor.fetchall()

model = whisper.load_model("medium")
repo_dir = config_data["repo_dir"]

for id, textline, clip_path in rows:
    print(f"Transcribing file {clip_path} with {textline} (row id: {id})...")
    result = model.transcribe(repo_dir + clip_path, language="Swedish")

    cleanTextLine = clean_string(textline)
    cleanTranscription = clean_string(result["text"])

    if (cleanTextLine == cleanTranscription):
        print(f"Approved : {cleanTranscription}");
        cursor.execute(f"UPDATE audio_clip SET status = 1 WHERE id = {id}")
        conn.commit()
    else:
        print(f"User required : {cleanTranscription}");
        cursor.execute(f"UPDATE audio_clip SET status = 2 WHERE id = {id}")
        conn.commit()

cursor.close()
conn.close()
